### ISSUE TITLE
nixos/mysql inconsistent dataDir setting

### ISSUE THREAD

**Issue by bjornfor** [NixOS/nixpkgs#17998]
## Issue description

mysql (mariadb) is built with `/var/lib/mysql` as datadir, but the default NixOS option for the mysql service is to run it with `/var/mysql` (no `lib/`). Because of this difference, using the recommended (most safe) way to reset root password[1] does not work:

```
$ sudo -u mysql mysqld_safe --init-file=/home/bfo/mysql-init
160825 22:33:08 mysqld_safe Logging to '/var/lib/mysql/mini.err'.
touch: cannot touch '/var/lib/mysql/mini.err': No such file or directory
chmod: cannot access '/var/lib/mysql/mini.err': No such file or directory
160825 22:33:08 mysqld_safe Starting mysqld daemon with databases from /var/lib/mysql
/run/current-system/sw/bin/mysqld_safe: line 144: /var/lib/mysql/mini.err: No such file or directory
/run/current-system/sw/bin/mysqld_safe: line 188: /var/lib/mysql/mini.err: No such file or directory
touch: cannot touch '/var/lib/mysql/mini.err': No such file or directory
chown: cannot access '/var/lib/mysql/mini.err': No such file or directory
chmod: cannot access '/var/lib/mysql/mini.err': No such file or directory
160825 22:33:08 mysqld_safe mysqld from pid file /var/lib/mysql/mini.pid ended
/run/current-system/sw/bin/mysqld_safe: line 144: /var/lib/mysql/mini.err: No such file or directory
```

`mysqld_safe --help` does not advertise a `--datadir` option that could have been used as workaround. So, either the user prepared for this and already set `services.mysql.dataDir = "/var/lib/mysql"` or (s)he would have to use the less secure and a bit more cumbersome password reset procedure.

I suggest that at some point the default datadir is set to `/var/lib/mysql`. Unfortunately, a breaking change.

[1] http://dev.mysql.com/doc/refman/5.7/en/resetting-permissions.html

**Comment by lancelotsix** [comment]
According to http://dev.mysql.com/doc/refman/5.7/en/mysqld-safe.html there is a `--datadir` switch available for `mysqld_safe`.

Anyway, `/var/lib/mysql` looks good to store mysql data.

**Comment by grahamc** [comment]
Meanwhile a  `TODO` in the mysql module says it should be put in to `/var/db/mysql`. My vote is for `/var/lib/mysql` :)

**Comment by bjornfor** [comment]
I thought moving datadir would be problamatic, but it looks like it is very easy and can be automated: http://stackoverflow.com/questions/1795176/how-to-change-mysql-data-directory.

Basically, stop mysql, move the datadir and possibly a ref in ```my.cnf``` and start mysql again.

**Comment by globin** [comment]
Can probably be done in `preStart`.

**Comment by grahamc** [comment]
Moving the datadir should likely come with a big warning to users, as there may be specific mounts in place.

**Comment by bjornfor** [comment]
Perhaps only (auto)move if old and new directory is on the same filesystem?

**Comment by grahamc** [comment]
IMHO the most safe and correct thing for us to do is:

1. On new systems (>= 17.09) use MySQL's default (/var/lib/mysql) datadir
2. On older systems, do nothing and leave the datadir in the same place it is using `stateVersion`

I feel:

1. trying to move it is dangerous and seriously playing with fire, think roll-backs and roll-forwards. Think datasets in the hundreds of gigabytes. Think other, out-of-band automation. Think systemd services with special access grants. Think apparmor profiles.

2. Making it match mongodb / postgresql is a foolish consistency, and moving to match upstream is a better choice.

**Comment by bjornfor** [comment]
Fixed with f469bd83bb9390807f1cdbe2de8d56bb8a710eb5 ("mysql service: change default data directory for 17.09"). Thanks!
